DESIGN.md

```diff
@@ -0,0 +1,102 @@
+# TeleWorker design document #
+
+- [Library](#library)
+- [Server](#server)
+- [Client](#client)
+- [Security](#security)
+- [Testing](#testing)
+- [Trade-offs](#trade-offs)
+
+## Library
+
+The library will contain all the functionality for starting, terminating, getting status, and streaming output of some jobs. The implementation can be split into 3 modules:
+
+1. Core - the main logic for jobs management.
+1. Broker - as we want to be able to stream some task output we need to provide all the logs produced by the task before the user connected as well as send all the upcoming logs.
+This can be handled by using os.Pipe() and a simple custom broker. Using a pipe we basically publish job stdout to a channel - there always will be at least one subscriber reading that channel, in such a way we can have a buffer with all the logs the job produced.
+Later, in case if the user wants to get the output - all we need to do is to create a copy of that buffer, subscribe so the copy could be also always updated with the latest logs in real time and simply stream that buffer.   
+1. Resource control - since the user is able to limit the task resources we need to have a small layer for working with the file system.  
+
+### Resource control
+Three subsystems will be used for allowing a user to limit a job:
+1.  blkio - parameter _weight_ is used.
+1.  cpu - parameter _shares_ is used. 
+1. memory - parameter _limit_in_bytes_ is used. It sets the upper limit of memory available to a particular job. 
+
+Initially, we set up a **teleworker** group with blkio.weight and cpu.shares parameters set to 1000 (i.e. maximum).
+Every job launched without limiting any of the parameters will be placed directly to **cgroup.procs** of that group, so basically we will write the job PID into the following:
```

Comment:
Can you provide some more details about your process execution model?

Reply:
@spirifoxy Basically, add some details on how your job worker server will start a job, make sure it's in a cgroup, etc.

Reply:
Added an overview of task creation flow to the Server section of the doc